### PR TITLE
Adding the Tfhe-rs-bool dialect 

### PR DESCRIPTION
**Pull request by WoutLegiest**
The previous tfhe-rs boolean operations and attributes are removed from the TfheRust dialect and putted in a separate dialect. All the tests (mlir + e2e) are changed accordingly.

### CODE REVIEW

File: lib/Target/TfheRustBool/TfheRustBoolEmitter.cpp
Comment (j2kun) [inline lib/Target/TfheRustBool/TfheRustBoolEmitter.cpp:1]: Would you be open to refactoring this so that the shared logic with TfheRustEmitter.cpp are unified in one place? OK to file as a followup ticket if you're not sure how to go about it.

File: tests/tfhe_rust_bool/end_to_end/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    #[arg(id = "input_2", index = 2, action)]
+    input2: u8,
+}
+
+fn main() {
+    let flags = Args::parse();
+    let (client_key, server_key) = tfhe::boolean::gen_keys();
+
+    let ct_1 = client_key.encrypt(flags.input1 == 0u8);
```
Comment (j2kun) [inline tests/tfhe_rust_bool/end_to_end/src/main.rs:20]: Seems like True should be equality with 1, not 0, right? I know it's a bug in the test harness, but maybe add another test to catch this since the existing test doesn't cover it (unlucky!).

File: tests/tfhe_rust_bool/types.mlir
Comment (j2kun) [inline tests/tfhe_rust_bool/types.mlir:1]: This file can probably be deleted since the same PR that introduces the type has other tests that exercise the type can be parsed correctly. We added this test in other cases where the ops were split into separate PRs from the types, and so this ensured the type-only PR was correct in the mean time.

**Review by j2kun — COMMENTED** [review]
A few nits, otherwise LGTM!

File: include/Dialect/TfheRust/IR/TfheRustOps.td [hunk 2]
```diff
@@ -17,26 +17,45 @@ class TfheRust_Op<string mnemonic, list<Trait> traits = []> :
...
+    TfheRust_ServerKey:$serverKey,
+    TfheRust_CiphertextType:$lhs,
+    TfheRust_CiphertextType:$rhs
+  );
+  let results = (outs TfheRust_CiphertextType:$output);
+}
 
-def CreateTrivialOp : TfheRust_Op<"create_trivial", [Pure]> {
-  let arguments = (ins TfheRust_ServerKey:$serverKey, AnyInteger:$value);
+def BitXorOp : TfheRust_Op<"bitxor", [
```
[the remark this answers is not in the document]
Comment (WoutLegiest) [reply include/Dialect/TfheRust/IR/TfheRustOps.td:50]: Now we rely on the tfhe api itself, making it somewhat more future proof? I think the maintenance this way is handled by zama

File: lib/Target/TfheRustBool/TfheRustBoolEmitter.cpp
Comment (WoutLegiest) [reply lib/Target/TfheRustBool/TfheRustBoolEmitter.cpp:1]: Definitely, that was on my planning as well. This is a lot of the same code

File: tests/tfhe_rust_bool/end_to_end/src/main.rs (hunk 1 above)
Comment (WoutLegiest) [reply tests/tfhe_rust_bool/end_to_end/src/main.rs:20]: I'll add it. It was done because the parser cannot (easily) process Boolean values

File: tests/tfhe_rust_bool/types.mlir
Comment (WoutLegiest) [reply tests/tfhe_rust_bool/types.mlir:1]: I'll remove it.

File: include/Dialect/TfheRust/IR/TfheRustOps.td (hunk 2 above)
[the remark this answers is not in the document]
Comment (j2kun) [reply include/Dialect/TfheRust/IR/TfheRustOps.td:50]: It's true, but the way I see it we don't need to add the ops until they're actually used in the codebase somewhere (either by passes generating them or in tests). It's just the YAGNI principle: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

It also happens that someone comes along, sees those ops, and thinks they are used when they are not, and then gets confused by doing some unnecessary work based on that false assumption.

[the remark this answers is not in the document]
Comment (WoutLegiest) [reply include/Dialect/TfheRust/IR/TfheRustOps.td]: Okay, principle makes sense in such a large codebase. I changed the file. I left the bitand, since the emitter has the output code

**Review by j2kun — CHANGES_REQUESTED** [review]
LGTM! As a final step, please manually squash the commits in this PR. Merging the imported CL inside Google will result in the commits preserved; there's sadly no way yet to automatically squash and merge.
